core/src/main/java/com/linecorp/armeria/common/Flags.java

```diff
@@ -1422,87 +1268,71 @@ private static ExceptionVerbosity exceptionLoggingMode(String name, String defau
+        if (value != null) {
+            value = Ascii.toLowerCase(value);
         }
 
-        if (validator.test(value)) {
-            logger.info("{}: {}", fullName, value);
-            return value;
+        if (value != null) {
+            if (validator.test(value)) {
+                logger.info("{}: {} (jvm option)", fullName, value);
+                return value;
+            }
+            logger.warn("{}: {} (jvm option) fail validation", fullName, value);
         }
-
-        logger.info("{}: {} (default instead of: {})", fullName, defaultValue, value);
+        logger.info("{}: {} (default)", fullName, defaultValue);
         return defaultValue;
     }
 
-    private static String getNormalized(String name, String defaultValue, Predicate<String> validator) {
-        final String fullName = PREFIX + name;
-        final String value = getLowerCased(fullName);
-        if (value == null) {
-            logger.info("{}: {} (default)", fullName, defaultValue);
-            return defaultValue;
-        }
+    private static <T> T getValue(Function<FlagsProvider, @Nullable T> method, String flagName) {
+        return getValue(method, flagName, unused -> true);
+    }
 
-        if (validator.test(value)) {
-            logger.info("{}: {}", fullName, value);
-            return value;
-        }
+    private static <T> T getValue(Function<FlagsProvider, @Nullable T> method,
+                                  String flagName, Predicate<T> validator) {
+        for (FlagsProvider provider : FLAGS_PROVIDERS) {
+            try {
+                final T value = method.apply(provider);
+                if (provider instanceof DefaultFlagsProvider) {
+                    logger.info("{}: {} ({})", flagName, value, provider.getClass().getSimpleName());
+                    return value;
+                }
 
-        logger.info("{}: {} (default instead of: {})", fullName, defaultValue, value);
-        return defaultValue;
-    }
+                if (value == null) {
+                    continue;
+                }
 
-    @Nullable
-    private static String getLowerCased(String fullName) {
-        String value = System.getProperty(fullName);
-        if (value != null) {
-            value = Ascii.toLowerCase(value);
+                if (!validator.test(value)) {
+                    logger.warn("{}: {} ({}) fail validation",
+                                flagName, value, provider.getClass().getSimpleName());
+                    continue;
+                }
+
+                logger.info("{}: {} ({})", flagName, value, provider.getClass().getSimpleName());
+                return value;
+            } catch (Exception ex) {
+                logger.warn("{}: ({}) fail to get value, {}", flagName, provider.getClass().getSimpleName(),
+                            ex.getMessage());
+            }
         }
-        return value;
+        throw new IllegalStateException(String.format("Cannot initial flag value : %s", flagName));
```

Comment:
How about returning the default value if there is an exception?
```suggestion
        return method.apply(DefaultFlagsProvider.INSTANCE);
```

Reply:
IllegalStateException can be thrown only in the case that DefaultFlagsProvider throws an exception when calling `method.apply(provider)` so even we change this line to return the default value. I think the exception will be thrown again. So I think throwing `IllegalStateException` might be useful to other contributors when they add/update an invalid default value during development. If I misunderstand something, please let me know, Thanks 🙏

Reply:
I understood. An `IllegalStateException` should not be raised in normal cases.
We usually choose either one of:
- Raise an exception if a receiver is able to handle the exception.
- Warn an exception as the last resort if it can't propagate to a caller or another context.
Since we 1) warn thrown exceptions and the `DefaultFlagsProvider` never throws an exception, we should never reach here.
How about raising an `Error` instead of `IllegalStateException`?
```java
// Should never reach here because DefaultFlagsProvider always returns a normal value.
throw new Error();
```

Reply:
I got it. I'll follow your suggestion. Thank you for the explanation 🙏.